File: app/src/main/java/com/rolandvitezhu/todocloud/MainActivity.java
package com.rolandvitezhu.todocloud;

import android.content.Context;
import android.os.Bundle;
import android.os.IBinder;
import android.support.annotation.NonNull;
import android.support.design.widget.CoordinatorLayout;
import android.support.design.widget.NavigationView;
import android.support.design.widget.Snackbar;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentTransaction;
import android.support.v4.view.GravityCompat;
import android.support.v4.widget.DrawerLayout;
import android.support.v7.app.ActionBar;
import android.support.v7.app.ActionBarDrawerToggle;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.view.ActionMode;
import android.support.v7.widget.Toolbar;
import android.view.MenuItem;
import android.view.View;
import android.view.inputmethod.InputMethodManager;
import android.widget.FrameLayout;
import android.widget.TextView;

import com.rolandvitezhu.todocloud.app.AppController;
import com.rolandvitezhu.todocloud.data.List;
import com.rolandvitezhu.todocloud.data.PredefinedList;
import com.rolandvitezhu.todocloud.data.Todo;
import com.rolandvitezhu.todocloud.data.User;
import com.rolandvitezhu.todocloud.datastorage.DbLoader;
import com.rolandvitezhu.todocloud.fragment.CreateTodoFragment;
import com.rolandvitezhu.todocloud.fragment.LoginUserFragment;
import com.rolandvitezhu.todocloud.fragment.LogoutUserDialogFragment;
import com.rolandvitezhu.todocloud.fragment.MainListFragment;
import com.rolandvitezhu.todocloud.fragment.ModifyPasswordFragment;
import com.rolandvitezhu.todocloud.fragment.ModifyTodoFragment;
import com.rolandvitezhu.todocloud.fragment.RegisterUserFragment;
import com.rolandvitezhu.todocloud.fragment.ResetPasswordFragment;
import com.rolandvitezhu.todocloud.fragment.SearchFragment;
import com.rolandvitezhu.todocloud.fragment.SettingsPreferenceFragment;
import com.rolandvitezhu.todocloud.fragment.TodoListFragment;
import com.rolandvitezhu.todocloud.helper.SessionManager;
import com.rolandvitezhu.todocloud.receiver.ReminderSetter;

import java.util.ArrayList;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;

public class MainActivity extends AppCompatActivity implements
    MainListFragment.IMainListFragment,
    LoginUserFragment.ILoginUserFragment,
    RegisterUserFragment.IRegisterUserFragment,
    ModifyPasswordFragment.IModifyPasswordFragment,
    ResetPasswordFragment.IResetPasswordFragment,
    FragmentManager.OnBackStackChangedListener,
    TodoListFragment.ITodoListFragment,
    ModifyTodoFragment.IModifyTodoFragmentActionBar,
    CreateTodoFragment.ICreateTodoFragmentActionBar,
    SettingsPreferenceFragment.ISettingsPreferenceFragment,
    LogoutUserDialogFragment.ILogoutUserDialogFragment,
    SearchFragment.ISearchFragment {

  @Inject
  DbLoader dbLoader;
  @Inject
  SessionManager sessionManager;

  @BindView(R.id.toolbar_main)
  Toolbar toolbar;
  @BindView(R.id.mainlist_drawerlayout)
  DrawerLayout drawerLayout;
  @BindView(R.id.mainlist_navigationview)
  NavigationView navigationView;
  @BindView(R.id.framelayout_main)
  FrameLayout container;
  @BindView(R.id.main_coordinator_layout)
  CoordinatorLayout coordinatorLayout;

  private ActionBarDrawerToggle actionBarDrawerToggle;

  @Override
  public void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    setContentView(R.layout.activity_main);
    ButterKnife.bind(this);

    ((AppController) getApplication()).getAppComponent().inject(this);

    setSupportActionBar(toolbar);

    prepareNavigationView(navigationView, toolbar);

    if (container != null) {
      if (savedInstanceState != null) {
        // Prevent Fragment overlapping
        return;
      }

      if (sessionManager.isLoggedIn()) {

        long id = getIntent().getLongExtra("id", -1);
        boolean wasMainActivityStartedFromLauncherIcon = id == -1;
        if (wasMainActivityStartedFromLauncherIcon) {
          openMainListFragment();
        } else if (wasMainActivityStartedFromNotification(id)){
          openMainListFragment();
          TodoListFragment todoListFragment = new TodoListFragment();
          openAllPredefinedList(todoListFragment);
          Todo notificationRelatedTodo = getNotificationRelatedTodo(id);
          openModifyTodoFragment(notificationRelatedTodo, todoListFragment);
        }
      } else {
        openLoginUserFragment();
      }

      prepareActionBarNavigationHandler();
      shouldDisplayHomeAsUp();
    }
  }

  private boolean wasMainActivityStartedFromNotification(long id) {
    return id != -1;
  }

  private void openAllPredefinedList(TodoListFragment todoListFragment) {
    String allPredefinedListWhere = dbLoader.prepareAllPredefinedListWhere();
    Bundle arguments = new Bundle();
    arguments.putString("selectFromDB", allPredefinedListWhere);
    arguments.putString("title", "2");
    arguments.putBoolean("isPredefinedList", true);
    todoListFragment.setArguments(arguments);
    openTodoListFragment(todoListFragment);
  }

  private Todo getNotificationRelatedTodo(long id) {
    return dbLoader.getTodo(id);
  }

  private void prepareActionBarNavigationHandler() {
    FragmentManager fragmentManager = getSupportFragmentManager();
    fragmentManager.addOnBackStackChangedListener(this);
  }

  private void shouldDisplayHomeAsUp() {
    FragmentManager fragmentManager = getSupportFragmentManager();
    boolean shouldDisplay = fragmentManager.getBackStackEntryCount()>0;
    ActionBar actionBar = getSupportActionBar();
    if (actionBar != null) {
      if (shouldDisplay) {
        setDrawerEnabled(false);
        actionBar.setDisplayHomeAsUpEnabled(true);
      } else {
        actionBar.setDisplayHomeAsUpEnabled(false);
        setDrawerEnabled(true);
      }

      CharSequence actionBarTitle = actionBar.getTitle();
      if (isMainListFragment(shouldDisplay, actionBarTitle)) {
        actionBar.setTitle(R.string.app_name);
        actionBar.setDisplayHomeAsUpEnabled(false);
        setDrawerEnabled(true);
      }
    }
  }

  private boolean isMainListFragment(boolean shouldDisplay, CharSequence actionBarTitle) {
    return !shouldDisplay && actionBarTitle != null && !actionBarTitle.equals(getString(R.string.all_login));
  }

  private void setDrawerEnabled(boolean enabled) {
    if (!enabled) {
      disableDrawer();
      enableActionBarBackNavigation();
    } else {
      enableDrawer();
    }
  }

  private void enableActionBarBackNavigation() {
    actionBarDrawerToggle.setToolbarNavigationClickListener(new View.OnClickListener() {

      @Override
      public void onClick(View view) {
        onBackPressed();
        hideSoftInput();
      }

    });
  }

  private void enableDrawer() {
    drawerLayout.setDrawerLockMode(DrawerLayout.LOCK_MODE_UNLOCKED);
    actionBarDrawerToggle.setDrawerIndicatorEnabled(true);
    actionBarDrawerToggle.syncState();
  }

  private void disableDrawer() {
    drawerLayout.setDrawerLockMode(DrawerLayout.LOCK_MODE_LOCKED_CLOSED);
    actionBarDrawerToggle.setDrawerIndicatorEnabled(false);
    actionBarDrawerToggle.syncState();
  }

  private void prepareNavigationView(NavigationView navigationView, Toolbar toolbar) {
    navigationView.setNavigationItemSelectedListener(
        new NavigationView.OnNavigationItemSelectedListener() {

          @Override
          public boolean onNavigationItemSelected(@NonNull MenuItem menuItem) {
            int menuItemId = menuItem.getItemId();

            switch (menuItemId) {
              case R.id.menuitem_navigationdrawer_settings:
                openSettingsPreferenceFragment();
                break;
              case R.id.menuitem_navigationdrawer_logout:
                openLogoutUserDialogFragment();
                break;
            }

            drawerLayout.closeDrawers();

            return true;
          }

        }
    );

    actionBarDrawerToggle = new ActionBarDrawerToggle(
        this,
        drawerLayout,
        toolbar,
        R.string.actionbardrawertoggle_opendrawer,
        R.string.actionbardrawertoggle_closedrawer
    );
    drawerLayout.addDrawerListener(actionBarDrawerToggle);
    actionBarDrawerToggle.syncState();
  }

  private void openLogoutUserDialogFragment() {
    LogoutUserDialogFragment logoutUserDialogFragment = new LogoutUserDialogFragment();
    logoutUserDialogFragment.show(getSupportFragmentManager(), "LogoutUserDialogFragment");
  }

  @Override
  public void onPrepareNavigationHeader() {
    View navigationHeader = navigationView.getHeaderView(0);
    TextView tvName = navigationHeader.findViewById(R.id.textview_navigationdrawerheader_name);
    TextView tvEmail = navigationHeader.findViewById(R.id.textview_navigationdrawerheader_email);
    User user = dbLoader.getUser();
    if (user != null) {
      tvName.setText(user.getName());
      tvEmail.setText(user.getEmail());
    }
  }

  @Override
  public void onSearchActionItemClick() {
    openSearchFragment();
  }

  private void openSearchFragment() {
    Bundle arguments = new Bundle(); // Will store UpdateTodoAdapter arguments later
    SearchFragment searchFragment = new SearchFragment();
    searchFragment.setArguments(arguments);
    FragmentManager fragmentManager = getSupportFragmentManager();
    FragmentTransaction fragmentTransaction = fragmentManager.beginTransaction();
    fragmentTransaction.replace(container.getId(), searchFragment);
    fragmentTransaction.addToBackStack(null);
    fragmentTransaction.commit();
  }

  @Override
  public boolean onOptionsItemSelected(MenuItem item) {
    int itemId = item.getItemId();

    switch (itemId) {
      case android.R.id.home:
        hideSoftInput();
        onBackPressed();
        break;
    }

    return super.onOptionsItemSelected(item);
  }

  private void hideSoftInput() {
    InputMethodManager inputMethodManager = (InputMethodManager) getSystemService(
        Context.INPUT_METHOD_SERVICE
    );
    View currentlyFocusedView = getCurrentFocus();
    if (currentlyFocusedView != null) {
      IBinder windowToken = currentlyFocusedView.getWindowToken();
      inputMethodManager.hideSoftInputFromWindow(windowToken, 0);
    }
  }

  @Override
  public void onBackPressed() {
    FragmentManager fragmentManager = getSupportFragmentManager();
    ModifyTodoFragment modifyTodoFragment = (ModifyTodoFragment) fragmentManager.findFragmentByTag(
        "ModifyTodoFragment"
    );
    if (modifyTodoFragment != null && modifyTodoFragment.isShouldNavigateBack()) {
      super.onBackPressed();
    } else if (modifyTodoFragment != null) {
      modifyTodoFragment.handleModifyTodo();
    } else if (fragmentManager.findFragmentByTag("RegisterUserFragment") != null) {
      navigateBackToLoginUserFragment();
    } else if (fragmentManager.findFragmentByTag("ResetPasswordFragment") != null) {
      navigateBackToLoginUserFragment();
    } else if (drawerLayout.isDrawerOpen(GravityCompat.START)) {
      drawerLayout.closeDrawers();
    } else {
      super.onBackPressed();
    }
  }

  private void navigateBackToLoginUserFragment() {
    super.onBackPressed();
    onSetActionBarTitle(getString(R.string.all_login));
  }

  @Override
  public void onClickPredefinedList(PredefinedList predefinedList) {
    Bundle arguments = new Bundle();
    arguments.putString("selectFromDB", predefinedList.getSelectFromDB());
    arguments.putString("title", predefinedList.getTitle());
    arguments.putBoolean("isPredefinedList", true);
    openTodoListFragment(arguments);
  }

  @Override
  public void onClickList(List list) {
    Bundle arguments = new Bundle();
    arguments.putString("listOnlineId", list.getListOnlineId());
    arguments.putString("title", list.getTitle());
    openTodoListFragment(arguments);
  }

  private void openTodoListFragment(Bundle arguments) {
    TodoListFragment todoListFragment = new TodoListFragment();
    todoListFragment.setArguments(arguments);
    FragmentManager fragmentManager = getSupportFragmentManager();
    FragmentTransaction fragmentTransaction = fragmentManager.beginTransaction();
    fragmentTransaction.replace(container.getId(), todoListFragment);
    fragmentTransaction.addToBackStack(null);
    fragmentTransaction.commit();
  }

  private void openTodoListFragment(TodoListFragment todoListFragment) {
    FragmentManager fragmentManager = getSupportFragmentManager();
    FragmentTransaction fragmentTransaction = fragmentManager.beginTransaction();
    fragmentTransaction.replace(container.getId(), todoListFragment);
    fragmentTransaction.addToBackStack(null);
    fragmentTransaction.commit();
  }

  @Override
  public void onLogout() {
    cancelReminders();
    sessionManager.setLogin(false);
    FragmentManager fragmentManager = getSupportFragmentManager();
    fragmentManager.popBackStackImmediate(null, FragmentManager.POP_BACK_STACK_INCLUSIVE);
    openLoginUserFragment();
    dbLoader.reCreateDb();
  }

  private void cancelReminders() {
    ArrayList<Todo> todosWithReminder = dbLoader.getTodosWithReminder();
    for (Todo todoWithReminder:todosWithReminder) {
      ReminderSetter.cancelReminderService(todoWithReminder);
    }
  }

  @Override
  public void onClickLinkToRegisterUser() {
    openRegisterUserFragment();
  }

  @Override
  public void onClickLinkToResetPassword() {
    openResetPasswordFragment();
  }

  private void openResetPasswordFragment() {
    ResetPasswordFragment resetPasswordFragment = new ResetPasswordFragment();
    FragmentManager fragmentManager = getSupportFragmentManager();
    FragmentTransaction fragmentTransaction = fragmentManager.beginTransaction();
    fragmentTransaction.replace(
        container.getId(),
        resetPasswordFragment,
        "ResetPasswordFragment"
    );
    fragmentTransaction.addToBackStack(null);
    fragmentTransaction.commit();
  }

  private void openRegisterUserFragment() {
    RegisterUserFragment registerUserFragment = new RegisterUserFragment();
    FragmentManager fragmentManager = getSupportFragmentManager();
    FragmentTransaction fragmentTransaction = fragmentManager.beginTransaction();
    fragmentTransaction.replace(
        container.getId(),
        registerUserFragment,
        "RegisterUserFragment"
    );
    fragmentTransaction.addToBackStack(null);
    fragmentTransaction.commit();
  }

  @Override
  public void onFinishRegisterUser() {
    FragmentManager fragmentManager = getSupportFragmentManager();
    fragmentManager.popBackStackImmediate(null, FragmentManager.POP_BACK_STACK_INCLUSIVE);
    openLoginUserFragment();
    onSetActionBarTitle(getString(R.string.all_login));
  }

  private void openLoginUserFragment() {
    LoginUserFragment loginUserFragment = new LoginUserFragment();
    FragmentManager fragmentManager = getSupportFragmentManager();
    FragmentTransaction fragmentTransaction = fragmentManager.beginTransaction();
    fragmentTransaction.replace(container.getId(), loginUserFragment);
    fragmentTransaction.commit();
  }

  @Override
  public void onFinishLoginUser() {
    FragmentManager fragmentManager = getSupportFragmentManager();
    fragmentManager.popBackStackImmediate(null, FragmentManager.POP_BACK_STACK_INCLUSIVE);
    openMainListFragment();
    onSetActionBarTitle(getString(R.string.app_name));
  }

  private void openMainListFragment() {
    MainListFragment mainListFragment = new MainListFragment();
    FragmentManager fragmentManager = getSupportFragmentManager();
    FragmentTransaction fragmentTransaction = fragmentManager.beginTransaction();
    fragmentTransaction.replace(container.getId(), mainListFragment);
    fragmentTransaction.commit();
  }

  @Override
  public void onClickChangePassword() {
    openModifyPasswordFragment();
  }

  private void openModifyPasswordFragment() {
    ModifyPasswordFragment modifyPasswordFragment = new ModifyPasswordFragment();
    FragmentManager fragmentManager = getSupportFragmentManager();
    FragmentTransaction fragmentTransaction = fragmentManager.beginTransaction();
    fragmentTransaction.replace(
        container.getId(),
        modifyPasswordFragment,
        "ModifyPasswordFragment"
    );
    fragmentTransaction.addToBackStack(null);
    fragmentTransaction.commit();
  }

  @Override
  public void onBackStackChanged() {
    shouldDisplayHomeAsUp();
  }

  @Override
  public boolean onSupportNavigateUp() {
    getSupportFragmentManager().popBackStack();
    return true;
  }

  @Override
  public void onFinishModifyPassword() {
    if (coordinatorLayout != null) {
      Snackbar snackbar = Snackbar.make(
          coordinatorLayout,
          R.string.modifypassword_passwordchangedsuccessfully,
          Snackbar.LENGTH_LONG
      );
      AppController.showWhiteTextSnackbar(snackbar);
    }
  }

  @Override
  public void onFinishResetPassword() {
    if (coordinatorLayout != null) {
      Snackbar snackbar = Snackbar.make(
          coordinatorLayout,
          R.string.resetpassword_passwordresetsuccessful,
          Snackbar.LENGTH_LONG
      );
      AppController.showWhiteTextSnackbar(snackbar);
    }
  }

  @Override
  public void onSetActionBarTitle(String title) {
    if (getSupportActionBar() != null)
      getSupportActionBar().setTitle(title);
    setDrawerEnabled(title.equals("Todo Cloud"));
  }

  @Override
  public void onStartActionMode(ActionMode.Callback callback) {
    startSupportActionMode(callback);
  }

  @Override
  public void onClickTodo(Todo todo, Fragment targetFragment) {
    openModifyTodoFragment(todo, targetFragment);
  }

  private void openModifyTodoFragment(Todo todo, Fragment targetFragment) {
    Bundle arguments = new Bundle();
    arguments.putParcelable("todo", todo);
    ModifyTodoFragment modifyTodoFragment = new ModifyTodoFragment();
    modifyTodoFragment.setTargetFragment(targetFragment, 0);
    modifyTodoFragment.setArguments(arguments);
    FragmentManager fragmentManager = getSupportFragmentManager();
    FragmentTransaction fragmentTransaction = fragmentManager.beginTransaction();
    fragmentTransaction.replace(
        container.getId(),
        modifyTodoFragment,
        "ModifyTodoFragment"
    );
    fragmentTransaction.addToBackStack(null);
    fragmentTransaction.commit();
  }

  @Override
  public void onOpenCreateTodoFragment(TodoListFragment targetFragment) {
    CreateTodoFragment createTodoFragment = new CreateTodoFragment();
    createTodoFragment.setTargetFragment(targetFragment, 0);
    FragmentManager fragmentManager = getSupportFragmentManager();
    FragmentTransaction fragmentTransaction = fragmentManager.beginTransaction();
    fragmentTransaction.replace(container.getId(), createTodoFragment);
    fragmentTransaction.addToBackStack(null);
    fragmentTransaction.commit();
  }

  public void openSettingsPreferenceFragment() {
    SettingsPreferenceFragment settingsPreferenceFragment = new SettingsPreferenceFragment();
    FragmentManager fragmentManager = getSupportFragmentManager();
    FragmentTransaction fragmentTransaction = fragmentManager.beginTransaction();
    fragmentTransaction.replace(container.getId(), settingsPreferenceFragment);
    fragmentTransaction.addToBackStack(null);
    fragmentTransaction.commit();
  }

}

File: app/src/main/java/com/rolandvitezhu/todocloud/ui/activity/main/MainActivity.kt
package com.rolandvitezhu.todocloud.ui.activity.main

import android.os.Bundle
import android.view.MenuItem
import android.view.View
import androidx.appcompat.app.ActionBarDrawerToggle
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.view.ActionMode
import androidx.core.view.GravityCompat
import androidx.databinding.DataBindingUtil
import androidx.drawerlayout.widget.DrawerLayout
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.lifecycleScope
import com.google.android.material.snackbar.Snackbar
import com.rolandvitezhu.todocloud.R
import com.rolandvitezhu.todocloud.app.AppController
import com.rolandvitezhu.todocloud.app.AppController.Companion.showWhiteTextSnackbar
import com.rolandvitezhu.todocloud.data.List
import com.rolandvitezhu.todocloud.data.PredefinedList
import com.rolandvitezhu.todocloud.data.Todo
import com.rolandvitezhu.todocloud.database.TodoCloudDatabase
import com.rolandvitezhu.todocloud.database.TodoCloudDatabaseDao
import com.rolandvitezhu.todocloud.databinding.ActivityMainBinding
import com.rolandvitezhu.todocloud.databinding.NavigationdrawerHeaderBinding
import com.rolandvitezhu.todocloud.helper.SessionManager
import com.rolandvitezhu.todocloud.helper.hideSoftInput
import com.rolandvitezhu.todocloud.receiver.ReminderSetter
import com.rolandvitezhu.todocloud.ui.activity.main.dialogfragment.LogoutUserDialogFragment
import com.rolandvitezhu.todocloud.ui.activity.main.fragment.*
import com.rolandvitezhu.todocloud.ui.activity.main.preferencefragment.SettingsPreferenceFragment
import com.rolandvitezhu.todocloud.ui.activity.main.viewmodel.ListsViewModel
import com.rolandvitezhu.todocloud.ui.activity.main.viewmodel.PredefinedListsViewModel
import com.rolandvitezhu.todocloud.ui.activity.main.viewmodel.TodosViewModel
import com.rolandvitezhu.todocloud.ui.activity.main.viewmodel.UserViewModel
import kotlinx.android.synthetic.main.activity_main.*
import kotlinx.android.synthetic.main.activity_main.view.*
import kotlinx.android.synthetic.main.layout_appbar.*
import kotlinx.android.synthetic.main.layout_appbar.view.*
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import javax.inject.Inject

class MainActivity : AppCompatActivity(), FragmentManager.OnBackStackChangedListener {

    @Inject
    lateinit var todoCloudDatabaseDao: TodoCloudDatabaseDao
    @Inject
    lateinit var todoCloudDatabase: TodoCloudDatabase
    @Inject
    lateinit var sessionManager: SessionManager

    private var actionBarDrawerToggle: ActionBarDrawerToggle? = null

    private val userViewModel by lazy {
        ViewModelProvider(this).get(UserViewModel::class.java)
    }
    private val todosViewModel by lazy {
        ViewModelProvider(this).get(TodosViewModel::class.java)
    }
    private val predefinedListsViewModel by lazy {
        ViewModelProvider(this).get(PredefinedListsViewModel::class.java)
    }
    private val listsViewModel by lazy {
        ViewModelProvider(this).get(ListsViewModel::class.java)
    }

    public override fun onCreate(savedInstanceState: Bundle?) {
        (application as AppController).appComponent.inject(this)
        super.onCreate(savedInstanceState)

        val activityMainBinding: ActivityMainBinding =
                DataBindingUtil.setContentView(this, R.layout.activity_main)
        val navigationdrawerHeaderViewBinding: NavigationdrawerHeaderBinding =
                NavigationdrawerHeaderBinding.inflate(
                        layoutInflater,
                        activityMainBinding.mainlistNavigationview,
                        false)
        activityMainBinding.mainlistNavigationview.addHeaderView(navigationdrawerHeaderViewBinding.root)
        val view: View = activityMainBinding.root

        activityMainBinding.mainActivity = this
        activityMainBinding.executePendingBindings()

        setSupportActionBar(view.toolbar_main)
        prepareNavigationView(view)

        navigationdrawerHeaderViewBinding.lifecycleOwner = this  // Set it, so as we update
        // the data in the userViewModel, the UI will update automatically.
        navigationdrawerHeaderViewBinding.userViewModel = userViewModel
        navigationdrawerHeaderViewBinding.executePendingBindings()

        if (view.framelayout_main != null) {
            if (savedInstanceState != null) {
                // Prevent Fragment overlapping
                return
            }
            if (sessionManager.isLoggedIn) {
                val id = intent.getLongExtra("id", -1)
                val wasMainActivityStartedFromLauncherIcon = id == -1L
                if (wasMainActivityStartedFromLauncherIcon) {
                    openMainListFragment(view)
                } else if (wasMainActivityStartedFromNotification(id)) {
                    lifecycleScope.launch {
                        todosViewModel.todo = getNotificationRelatedTodo(id)
                    }
                    openMainListFragment(view)
                    val todoListFragment = TodoListFragment()
                    openAllPredefinedList(todoListFragment, view)
                    openModifyTodoFragment(todoListFragment, view)
                }
            } else {
                openLoginUserFragment(view)
            }
            prepareActionBarNavigationHandler()
            shouldDisplayHomeAsUp(view)
        }
        ReminderSetter.createReminderServices(applicationContext)
    }

    private fun wasMainActivityStartedFromNotification(id: Long): Boolean {
        return id != -1L
    }

    private fun openAllPredefinedList(todoListFragment: TodoListFragment, view: View) {
        lifecycleScope.launch {
            val allPredefinedListWhere = todoCloudDatabaseDao.prepareAllPredefinedListWhere()
            val predefinedList = PredefinedList(getString(R.string.all_all), allPredefinedListWhere)
            predefinedListsViewModel.predefinedList = predefinedList
            todosViewModel.updateTodosViewModelByWhereCondition(
                    predefinedList.title,
                    predefinedList.selectFromDB
            )
        }
        openTodoListFragment(todoListFragment, view)
    }

    private suspend fun getNotificationRelatedTodo(id: Long): Todo {
        return todoCloudDatabaseDao.getTodo(id)?: Todo()
    }

    private fun prepareActionBarNavigationHandler() {
        val fragmentManager = supportFragmentManager
        fragmentManager.addOnBackStackChangedListener(this)
    }

    private fun shouldDisplayHomeAsUp(view: View?) {
        val fragmentManager = supportFragmentManager
        val shouldDisplay = fragmentManager.backStackEntryCount > 0
        val actionBar = supportActionBar
        if (actionBar != null) {
            if (shouldDisplay) {
                setDrawerEnabled(false, view)
                actionBar.setDisplayHomeAsUpEnabled(true)
            } else {
                actionBar.setDisplayHomeAsUpEnabled(false)
                setDrawerEnabled(true, view)
            }
            val actionBarTitle = actionBar.title
            if (isMainListFragment(shouldDisplay, actionBarTitle)) {
                actionBar.setTitle(R.string.app_name)
                actionBar.setDisplayHomeAsUpEnabled(false)
                setDrawerEnabled(true, view)
            }
        }
    }

    private fun isMainListFragment(shouldDisplay: Boolean, actionBarTitle: CharSequence?): Boolean {
        return !shouldDisplay && actionBarTitle != null && actionBarTitle != getString(R.string.all_login)
    }

    private fun setDrawerEnabled(enabled: Boolean, view: View?) {
        if (!enabled) {
            disableDrawer(view)
            enableActionBarBackNavigation()
        } else {
            enableDrawer(view)
        }
    }

    private fun enableActionBarBackNavigation() {
        actionBarDrawerToggle?.toolbarNavigationClickListener = View.OnClickListener {
            onBackPressed()
            hideSoftInput()
        }
    }

    private fun enableDrawer(view: View?) {
        if (view != null)
            view.mainlist_drawerlayout?.setDrawerLockMode(DrawerLayout.LOCK_MODE_UNLOCKED)
        else
            this.mainlist_drawerlayout?.setDrawerLockMode(DrawerLayout.LOCK_MODE_UNLOCKED)
        actionBarDrawerToggle?.isDrawerIndicatorEnabled = true
        actionBarDrawerToggle?.syncState()
    }

    private fun disableDrawer(view: View?) {
        if (view != null)
            view.mainlist_drawerlayout?.setDrawerLockMode(DrawerLayout.LOCK_MODE_LOCKED_CLOSED)
        else
            this.mainlist_drawerlayout?.setDrawerLockMode(DrawerLayout.LOCK_MODE_LOCKED_CLOSED)
        actionBarDrawerToggle?.isDrawerIndicatorEnabled = false
        actionBarDrawerToggle?.syncState()
    }

    private fun prepareNavigationView(view: View) {
        view.mainlist_navigationview?.setNavigationItemSelectedListener { menuItem ->
            when (menuItem.itemId) {
                R.id.menuitem_navigationdrawer_settings -> openSettingsPreferenceFragment(view)
                R.id.menuitem_navigationdrawer_logout -> openLogoutUserDialogFragment()
            }
            view.mainlist_drawerlayout?.closeDrawers()
            true
        }
        actionBarDrawerToggle = ActionBarDrawerToggle(
                this,
                view.mainlist_drawerlayout,
                view.toolbar_main,
                R.string.actionbardrawertoggle_opendrawer,
                R.string.actionbardrawertoggle_closedrawer
        )
        actionBarDrawerToggle?.let { view.mainlist_drawerlayout?.addDrawerListener(it) }
        actionBarDrawerToggle?.syncState()
    }

    private fun openLogoutUserDialogFragment() {
        val logoutUserDialogFragment = LogoutUserDialogFragment()
        logoutUserDialogFragment.show(supportFragmentManager, "LogoutUserDialogFragment")
    }

    fun onPrepareNavigationHeader() {
        lifecycleScope.launch {
            userViewModel.updateUserViewModel()
        }
    }

    fun onSearchActionItemClick() {
        openSearchFragment()
    }

    private fun openSearchFragment() {
        todosViewModel.todos.value?.clear()
        val searchFragment = SearchFragment()
        val fragmentManager = supportFragmentManager
        val fragmentTransaction = fragmentManager.beginTransaction()
        fragmentTransaction.replace(this.framelayout_main!!.id, searchFragment)
        fragmentTransaction.addToBackStack(null)
        fragmentTransaction.commit()
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val itemId = item.itemId
        when (itemId) {
            android.R.id.home -> {
                hideSoftInput()
                onBackPressed()
            }
        }

        return super.onOptionsItemSelected(item)
    }

    override fun onBackPressed() {
        val fragmentManager = supportFragmentManager
        val modifyTodoFragment = fragmentManager.findFragmentByTag(
                "ModifyTodoFragment"
        ) as ModifyTodoFragment?
        if (modifyTodoFragment != null && todosViewModel.shouldNavigateBack) {
            super.onBackPressed()
        } else if (modifyTodoFragment != null) {
            modifyTodoFragment.handleModifyTodo()
        } else if (fragmentManager.findFragmentByTag("RegisterUserFragment") != null) {
            navigateBackToLoginUserFragment()
        } else if (fragmentManager.findFragmentByTag("ResetPasswordFragment") != null) {
            navigateBackToLoginUserFragment()
        } else if (this.mainlist_drawerlayout?.isDrawerOpen(GravityCompat.START) == true) {
            this.mainlist_drawerlayout?.closeDrawers()
        } else {
            super.onBackPressed()
        }
    }

    private fun navigateBackToLoginUserFragment() {
        super.onBackPressed()
        onSetActionBarTitle(getString(R.string.all_login))
    }

    fun onClickPredefinedList(predefinedList: PredefinedList) {
        lifecycleScope.launch {
            todosViewModel.updateTodosViewModelByWhereCondition(
                    predefinedList.title,
                    predefinedList.selectFromDB
            )
        }
        openTodoListFragment()
    }

    fun onClickList(list: List) {
        lifecycleScope.launch {
            list.listOnlineId?.let {
                todosViewModel.updateTodosViewModelByListOnlineId(
                        list.title,
                        it
                )
            }
        }
        openTodoListFragment()
    }

    private fun openTodoListFragment() {
        val todoListFragment = TodoListFragment()
        val fragmentManager = supportFragmentManager
        val fragmentTransaction = fragmentManager.beginTransaction()
        fragmentTransaction.replace(this.framelayout_main!!.id, todoListFragment)
        fragmentTransaction.addToBackStack(null)
        fragmentTransaction.commit()
    }

    private fun openTodoListFragment(todoListFragment: TodoListFragment, view: View) {
        val fragmentManager = supportFragmentManager
        val fragmentTransaction = fragmentManager.beginTransaction()
        fragmentTransaction.replace(view.framelayout_main!!.id, todoListFragment)
        fragmentTransaction.addToBackStack(null)
        fragmentTransaction.commit()
    }

    fun onLogout() {
        cancelReminders()
        sessionManager.setLogin(false)
        val fragmentManager = supportFragmentManager
        fragmentManager.popBackStackImmediate(null, FragmentManager.POP_BACK_STACK_INCLUSIVE)
        openLoginUserFragment(null)
        GlobalScope.launch(Dispatchers.IO) { todoCloudDatabase.clearAllTables() }
    }

    private fun cancelReminders() {
        lifecycleScope.launch {
            val todosWithReminder = todoCloudDatabaseDao.getTodosWithReminder()
            for (todoWithReminder in todosWithReminder) {
                ReminderSetter.cancelReminderService(todoWithReminder)
            }
        }
    }

    fun onClickLinkToRegisterUser() {
        openRegisterUserFragment()
    }

    fun onClickLinkToResetPassword() {
        openResetPasswordFragment()
    }

    private fun openResetPasswordFragment() {
        val resetPasswordFragment = ResetPasswordFragment()
        val fragmentManager = supportFragmentManager
        val fragmentTransaction = fragmentManager.beginTransaction()
        fragmentTransaction.replace(
                this.framelayout_main!!.id,
                resetPasswordFragment,
                "ResetPasswordFragment"
        )
        fragmentTransaction.addToBackStack(null)
        fragmentTransaction.commit()
    }

    private fun openRegisterUserFragment() {
        val registerUserFragment = RegisterUserFragment()
        val fragmentManager = supportFragmentManager
        val fragmentTransaction = fragmentManager.beginTransaction()
        fragmentTransaction.replace(
                this.framelayout_main!!.id,
                registerUserFragment,
                "RegisterUserFragment"
        )
        fragmentTransaction.addToBackStack(null)
        fragmentTransaction.commit()
    }

    fun onFinishRegisterUser() {
        val fragmentManager = supportFragmentManager
        fragmentManager.popBackStackImmediate(null, FragmentManager.POP_BACK_STACK_INCLUSIVE)
        openLoginUserFragment(null)
        onSetActionBarTitle(getString(R.string.all_login))
    }

    private fun openLoginUserFragment(view: View?) {
        val loginUserFragment = LoginUserFragment()
        val fragmentManager = supportFragmentManager
        val fragmentTransaction = fragmentManager.beginTransaction()
        if (view != null)
            fragmentTransaction.replace(view.framelayout_main!!.id, loginUserFragment)
        else
            fragmentTransaction.replace(this.framelayout_main!!.id, loginUserFragment)
        fragmentTransaction.commit()
    }

    fun onFinishLoginUser() {
        val fragmentManager = supportFragmentManager
        fragmentManager.popBackStackImmediate(null, FragmentManager.POP_BACK_STACK_INCLUSIVE)
        openMainListFragment(null)
        onSetActionBarTitle(getString(R.string.app_name))
    }

    private fun openMainListFragment(view: View?) {
        val mainListFragment = MainListFragment()
        val fragmentManager = supportFragmentManager
        val fragmentTransaction = fragmentManager.beginTransaction()
        if (view != null)
            fragmentTransaction.replace(view.framelayout_main!!.id, mainListFragment)
        else
            fragmentTransaction.replace(this.framelayout_main!!.id, mainListFragment)
        fragmentTransaction.commit()
    }

    fun onClickChangePassword() {
        openModifyPasswordFragment()
    }

    private fun openModifyPasswordFragment() {
        val modifyPasswordFragment = ModifyPasswordFragment()
        val fragmentManager = supportFragmentManager
        val fragmentTransaction = fragmentManager.beginTransaction()
        fragmentTransaction.replace(
                this.framelayout_main!!.id,
                modifyPasswordFragment,
                "ModifyPasswordFragment"
        )
        fragmentTransaction.addToBackStack(null)
        fragmentTransaction.commit()
    }

    override fun onBackStackChanged() {
        shouldDisplayHomeAsUp(null)
    }

    override fun onSupportNavigateUp(): Boolean {
        supportFragmentManager.popBackStack()
        return true
    }

    fun onFinishModifyPassword() {
        try {
            val snackbar = Snackbar.make(
                    this.main_coordinator_layout!!,
                    R.string.modifypassword_passwordchangedsuccessfully,
                    Snackbar.LENGTH_LONG
            )
            showWhiteTextSnackbar(snackbar)
        } catch (e: NullPointerException) {
            // Snackbar or coordinatorLayout doesn't exists already.
        }
    }

    fun onFinishResetPassword() {
        try {
            val snackbar = Snackbar.make(
                    this.main_coordinator_layout!!,
                    R.string.resetpassword_passwordresetsuccessful,
                    Snackbar.LENGTH_LONG
            )
            showWhiteTextSnackbar(snackbar)
        } catch (e: NullPointerException) {
            // Snackbar or coordinatorLayout doesn't exists already.
        }
    }

    fun onSetActionBarTitle(title: String?) {
        supportActionBar?.title = title
        setDrawerEnabled(title == "Todo Cloud", null)
    }

    fun onStartActionMode(callback: ActionMode.Callback?) {
        callback?.let { startSupportActionMode(it) }
    }

    fun openModifyTodoFragment(targetFragment: Fragment?, view: View?) {
        val modifyTodoFragment = ModifyTodoFragment()
        modifyTodoFragment.setTargetFragment(targetFragment, 0)

        val fragmentTransaction = supportFragmentManager.beginTransaction()
        if (view != null)
            fragmentTransaction.replace(
                    view.framelayout_main.id,
                    modifyTodoFragment,
                    "ModifyTodoFragment"
            )
        else
            fragmentTransaction.replace(
                    this.framelayout_main.id,
                    modifyTodoFragment,
                    "ModifyTodoFragment"
            )
        fragmentTransaction.addToBackStack(null)
        fragmentTransaction.commit()
    }

    fun onOpenCreateTodoFragment(targetFragment: TodoListFragment?) {
        val createTodoFragment = CreateTodoFragment()
        createTodoFragment.setTargetFragment(targetFragment, 0)

        val fragmentTransaction = supportFragmentManager.beginTransaction()
        fragmentTransaction.replace(this.framelayout_main.id, createTodoFragment)
        fragmentTransaction.addToBackStack(null)
        fragmentTransaction.commit()
    }

    fun openSettingsPreferenceFragment(view: View) {
        val settingsPreferenceFragment = SettingsPreferenceFragment()

        val fragmentTransaction = supportFragmentManager.beginTransaction()
        fragmentTransaction.replace(view.framelayout_main.id, settingsPreferenceFragment)
        fragmentTransaction.addToBackStack(null)
        fragmentTransaction.commit()
    }

    fun onModifyTodo() {
        lifecycleScope.launch {
            todoCloudDatabaseDao.updateTodo(todosViewModel.todo)
            todoCloudDatabaseDao.fixTodoPositions()
            todosViewModel.updateTodosViewModel()
        }

        if (isSetReminder(todosViewModel.todo)) {
            if (shouldCreateReminderService(todosViewModel.todo)) {
                ReminderSetter.createReminderService(todosViewModel.todo)
            }
        } else {
            ReminderSetter.cancelReminderService(todosViewModel.todo)
        }
    }

    private fun isSetReminder(todo: Todo): Boolean {
        return !todo.reminderDateTime?.equals("-1")!!
    }

    private fun shouldCreateReminderService(todoToModify: Todo): Boolean {
        return isNotCompleted(todoToModify) && isNotDeleted(todoToModify)
    }

    private fun isNotCompleted(todo: Todo): Boolean {
        return todo.completed?.not() ?: true
    }

    private fun isNotDeleted(todo: Todo): Boolean {
        return todo.deleted?.not() ?: true
    }
}